app/services/storage_root_migration_service.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# Migrates Complete Moab records to a new Moab Storate Root.
+class StorageRootMigrationService
+  def initialize(from_name, to_name)
+    @from_name = from_name
+    @to_name = to_name
+  end
+
+  # @return [Array<String>] druids of migrated moabs
+  def migrate
+    druids = []
+    from_root.complete_moabs.find_each do |complete_moab|
+      migrate_moab(complete_moab)
+      druids << complete_moab.preserved_object.druid
+    end
```

Comment:
my recollection is that we said we'd only run this migration once
1. the contents of the original storage root had been fully copied to the destination storage root
2. writing of all new objects or versions of existing objects to the old storage root had stopped (i.e. we've paused accessioning for the weekend to do the cutover)

and i think we want to make that explicit in the relevant code and/or docs (so maybe comment here and in rake task description?).

Reply:
also i played around with an alternate implementation of this method that just did a one shot `update_all` (initially thought to try because i'd forgotten that accessioning would be turned off, and worried about `find_each` batching, but after i realized that was a non-issue, it still seemed like it might be a little more concise and a little more efficient)...

...but then i realized that skips ActiveRecord validations and hooks, and so CV jobs would have to be queued explicitly, and then i think that might lose the concision benefit, which would make that approach much less worhtwhile, imho.  maybe you considered all that already when implementing this and i was _totally_ wasting my time (as opposed to _sort of_ wasting it) 😆 

branch is here, though, if you're curious:  https://github.com/sul-dlss/preservation_catalog/compare/t1322-migrate_task...t1322-migrate_task-single_update?expand=1

Reply:
druids is going to be btw 90K - 150K records in that array per root; is that OK?

Reply:
i think that's fine, whether using `find_each` or `update_all`.  i expect PG to be able to update hundreds of thousands of records, esp if all other traffic is halted.  and while instantiating and chewing through 150K ActiveRecord objects (the `find_each` will instantiate a `CompleteMoab` for every row it retrieves) will prob be a little slow, i think it'll be tractable.

since we're planning on doing a dry run with prod data, i'm not too worried that we _have_ to figure this out ahead of time purely through speculation, and i'm pretty confident we can figure something out without too much trouble if our dry run ends up taking too long.  we could start with a smaller root for benchmarking.

FWIW, in working on #1317, i was able to dump the list of druids from the largest storage root in what felt like about 3-5 seconds.  that was just a select for the druids with no update, but it resulted in about 350K druids.

i would expect that the trouble would come if we tried to do an `update_all` _while_ the root was still actively being added to.  that sounds like it'd invite deadlock hell, but i don't think that's our plan here.

Reply:
also to make it explicit:  the skipping of AR validations by `update_all` didn't worry me in this particular case, because i think the FK constraint on `moab_storage_root_id` and the translation of the `status` string to its int code provide all the validation we need for this very limited set of updates.  might feel differently when updating a different set of fields.

Reply:
if people ended up liking it (see usage/discussion in #1347), this might be worth considering for a follow on PR:  https://github.com/afair/postgresql_cursor

pros:  seems like if that gem were in place, you could swap `find_each` for `each_instance` and then you'd be able to page over results and use AR objects with no other changes, and have it all happen inside a transaction.

cons:  that'd be a long running transaction, though since accessioning should be off, there shouldn't be contention for the rows, so maybe ok? 🤷‍♂   and i'd probably want to add a test to make sure the CV jobs got queued for the moved moabs, just to make sure the gem was instantiating the AR object properly (i.e. in a way that triggers hooks properly).

anyway, just a thought, definitely would not do that here.

Reply:
also maybe updates inside a cursor is a bad idea?  and PG docs specifically advise against select for update and row locking, iirc.